torch/distributed/pipelining/schedules.py

```diff
@@ -660,71 +805,78 @@ def _step_microbatches(
+                        raise ValueError(f"Unknown computation type {computation_type}")
+
+                # Look at the neighboring ranks for this current timestep and determine whether
+                # this current rank needs to do any recv communication
+                prev_rank_action = None
+                if time_step < len(prev_rank_ops):
+                    prev_rank_action = prev_rank_ops[time_step]
+                if prev_rank_action is not None:
+                    computation_type, mb_index, stage_index = prev_rank_action
+                    # Only handle sends for the forward from a previous rank
+                    if computation_type == _ComputationType.FORWARD:
+                        # If not the last stage, then receive fwd activations
+                        if stage_index != self._num_stages - 1:
+                            # TODO: We are assuming that stage will always receive from stage-1
+                            # however that is not necessarily true of get_fwd_recv_ops
+                            stage = stage_index_to_stage[stage_index + 1]
+                            ops.extend(stage.get_fwd_recv_ops(mb_index))
+                    elif computation_type == _ComputationType.BACKWARD:
+                        # Previous rank doing backward has no influence for the current rank forward recv
+                        pass
+                    else:
+                        raise ValueError(f"Unknown computation type {computation_type}")
+
+                next_rank_action = None
+                if time_step < len(next_rank_ops):
+                    next_rank_action = next_rank_ops[time_step]
+                if next_rank_action is not None:
+                    computation_type, mb_index, stage_index = next_rank_action
+                    # Only handle receives for the backwards from a next rank
+                    if computation_type == _ComputationType.FORWARD:
+                        # Next rank doing forward has no influence for the current rank backward recv
+                        pass
+                    elif computation_type == _ComputationType.BACKWARD:
+                        # If not the first stage, then receive bwd gradients
+                        if stage_index != 0:
+                            # TODO: We are assuming that stage will always receive from stage+1
+                            # however that is not necessarily true of get_bwd_recv_ops
+                            stage = stage_index_to_stage[stage_index - 1]
+                            ops.extend(stage.get_bwd_recv_ops(mb_index))
+                    else:
+                        raise ValueError(f"Unknown computation type {computation_type}")
+
+                # do the communication
+                if ops:
+                    _batch_p2p(ops).wait()
+            except Exception as e:
+                logger.error(
+                    "Exception in rank %s at time step %s", self.rank, time_step
```

Comment:
this string should include more info.  not sure what amount is right but some things that might be good 
 - which stage?
 - which action?

Maybe also phrase it more like "PipelineSchedule {schedulename?} caught the following exception when running {}th microbatch {Action} for stage {}" or something?